### PR TITLE
CB-15708 Add hbase virtual group to the property ranger.hbase.default…

### PR DESCRIPTION
**Pull request by ragarkar**
1. Added a new configuration ranger.hbase.default.admin.groups. This configuration will be used by ranger to associate the corresponding Hbase virtual admin group to the default "all - tables, column-family, columns" policy.
2. Added code to assign the HBase virtual admin group to the newly added configuration.
3. Added a unit test to test this change.

### CODE REVIEW

**Review by doktoric — APPROVED** [review]

**Comment by ragarkar** [comment]
recheck

**Comment by subrat-mishra** [comment]
recheck

**Comment by ragarkar** [comment]
recheck

File: template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/hbase/HbaseServiceConfigProviderTest.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import com.sequenceiq.cloudbreak.auth.altus.EntitlementService;
+import com.sequenceiq.cloudbreak.auth.altus.UmsRight;
+import com.sequenceiq.cloudbreak.auth.altus.VirtualGroupRequest;
+import com.sequenceiq.cloudbreak.auth.altus.VirtualGroupService;
+import com.sequenceiq.cloudbreak.cmtemplate.CmTemplateProcessor;
+import com.sequenceiq.cloudbreak.template.TemplatePreparationObject;
+import java.util.List;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
```
Comment (lajosrodek) [inline template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/hbase/HbaseServiceConfigProviderTest.java:17]: Please use JUnit 5 only; mixing annotations from JUnit 4 & 5 is not ideal.

File: template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/hbase/HbaseServiceConfigProviderTest.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
+package com.sequenceiq.cloudbreak.cmtemplate.configproviders.hbase;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.doReturn;
+import static org.junit.Assert.assertTrue;
```
Comment (lajosrodek) [inline template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/hbase/HbaseServiceConfigProviderTest.java:5]: Like below, please use JUnit 5 only consistently.

File: template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/hbase/HbaseServiceConfigProviderTest.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    @InjectMocks
+    HbaseServiceConfigProvider underTest;
+
+    @Mock
+    VirtualGroupService virtualGroupService;
+
+    @Before
+    public void setUp() {
+        MockitoAnnotations.initMocks(this);
+    }
```
Comment (lajosrodek) [inline template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/hbase/HbaseServiceConfigProviderTest.java:40]: Not needed,  `MockitoExtension` takes care of that.

File: template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/hbase/HbaseServiceConfigProviderTest.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    private static final String MOCK_HBASE_VIRTUAL_ADMIN_GROUP = "_c_hbase_admin";
+
+    private static final String RANGER_HBASE_ADMIN_VIRTUAL_GROUPS = "ranger.hbase.default.admin.groups";
+
+    @InjectMocks
+    HbaseServiceConfigProvider underTest;
+
+    @Mock
+    VirtualGroupService virtualGroupService;
```
Comment (lajosrodek) [inline template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/hbase/HbaseServiceConfigProviderTest.java:35]: Please make these `private`.

File: template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/hbase/HbaseServiceConfigProviderTest.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        CmTemplateProcessor cmTemplateProcessor = new CmTemplateProcessor(blueprintText);
+        cmTemplateProcessor.setCmVersion("7.6.0");
+        VirtualGroupRequest virtualGroupRequest = new VirtualGroupRequest(MOCK_CRN, MOCK_HBASE_VIRTUAL_ADMIN_GROUP);
+        TemplatePreparationObject templatePreparationObject = TemplatePreparationObject.Builder.builder()
+            .withVirtualGroupView(virtualGroupRequest)
+            .build();
+        doReturn(MOCK_HBASE_VIRTUAL_ADMIN_GROUP)
+            .when(virtualGroupService).getVirtualGroup(virtualGroupRequest, UmsRight.HBASE_ADMIN.getRight());
+        List<ApiClusterTemplateConfig> serviceConfigs = underTest.getServiceConfigs(cmTemplateProcessor, templatePreparationObject);
+        assertEquals("Unexpected number of configurations returned: ", 2, serviceConfigs.size());
```
Comment (lajosrodek) [inline template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/hbase/HbaseServiceConfigProviderTest.java:60]: While it does not strictly relate to the present commit, could you please also verify the correctness of `SPNEGO_ADMIN_GROUP`?

**Review by lajosrodek — CHANGES_REQUESTED** [review]

**Comment by sidseth** [comment]
@ragarkar - A few questions on the change?
This affects SDX cluster creation... does it also require clients to receive config changes?
What happens if there are older DH HBase clusters (let's say 7.2.11) running against a newer SDX (wherever this patch lands)?
... or maybe this has 0 impact on clients?

Should this change be in CM, and in the CM upgrade handlers instead of Cloudbreak?

**Comment by ragarkar** [comment]
recheck

**Comment by ragarkar** [comment]
recheck: Tests did not trigger on previous attempts.

**Comment by ragarkar** [comment]
> @ragarkar - A few questions on the change. This affects SDX cluster creation... does it also require clients to receive config changes? What happens if there are older DH HBase clusters (let's say 7.2.11) running against a newer SDX (wherever this patch lands)? ... or maybe this has 0 impact on clients?
=====> AFAIU, this change does not need any changes to the client. There are two parts to this fix. This change in CB introduces a new configuration variable which specifies the Hbase virtual admin group. This configuration variable in turn is read by the Ranger code in ODX-546 to associated the corresponding Hbase virtual admin group to the default policy. Since both the CB and the Ranger code will come into effect only when both the components have correct versions deployed, I don't think there is any issues with the CB change working with older clusters running against a newer SDX.
> 
> Should this change be in CM, and in the CM upgrade handlers instead of Cloudbreak?
=====> AFAIU, CM is not aware of any configuration. Instead, it uses the configuration exposed by the CB code and hence, I think, the change is needed in CB instead of CM. Please let me know if my understanding is wrong.

File: template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/hbase/HbaseServiceConfigProviderTest.java (hunk 1 above)
Comment (ragarkar) [reply template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/hbase/HbaseServiceConfigProviderTest.java]: Changed the code to use JUnit 5 annotations.

File: template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/hbase/HbaseServiceConfigProviderTest.java (hunk 3 above)
Comment (ragarkar) [reply template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/hbase/HbaseServiceConfigProviderTest.java]: Removed this code and verified the run.

File: template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/hbase/HbaseServiceConfigProviderTest.java (hunk 4 above)
Comment (ragarkar) [reply template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/hbase/HbaseServiceConfigProviderTest.java]: Done.

File: template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/hbase/HbaseServiceConfigProviderTest.java (hunk 5 above)
Comment (ragarkar) [reply template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/hbase/HbaseServiceConfigProviderTest.java]: Added checks for SPNEGO_ADMIN_GROUP.

**Comment by ragarkar** [comment]
Please do not merge as I am still doing the testing.

**Review by lajosrodek — APPROVED** [review]
Thanks for the changes! LGTM

**Comment by ragarkar** [comment]
Changed the approach after discussion with Rajesh. Changed the code to introduce the new config as a Ranger config rather than as Hbase config.

**Review by lajosrodek — APPROVED** [review]
Still looks OK with the new approach.

**Pull request merged**